File: src/main/java/com/wseemann/ecp/core/ECPRequest.java
package com.wseemann.ecp.core;

import java.io.IOException;

import com.wseemann.ecp.logging.Logger;
import com.wseemann.ecp.parser.ECPResponseParser;
import okhttp3.*;
import okhttp3.Request;
import org.jdom2.JDOMException;

public abstract class ECPRequest<T> {

	private static final String DISCOVERY = "DISCOVERY";

	private final String url;

	protected ECPRequest(String url) {
		this.url = url;
	}

	protected abstract String getMethod();
	protected abstract String getPath();

	protected abstract ECPResponseParser<T> getParser();

	public ECPResponse<T> send() throws IOException {
		String url = this.url + getPath();

		try {
			if (getMethod().equalsIgnoreCase(DISCOVERY)) {
				com.wseemann.ecp.core.Request request = new DiscoveryRequest(url);
				return new ECPResponse<>(generateResponseData(request.send().getData(), getParser()));
			} else {
				OkHttpClient.Builder okHttpBuilder = new OkHttpClient.Builder();
				okHttpBuilder.setConnectTimeout$okhttp(6000);
				okHttpBuilder.setReadTimeout$okhttp(6000);
				OkHttpClient okHttpClient = okHttpBuilder.build();

				Request.Builder okHttpRequestBuilder = new okhttp3.Request.Builder()
						.addHeader("User-Agent", "Roku-ECP-Wrapper-Kotlin");
				okHttpRequestBuilder.setUrl$okhttp(HttpUrl.parse(url));
				okHttpRequestBuilder.setMethod$okhttp(getMethod());
				Request request = okHttpRequestBuilder.build();

				Call call = okHttpClient.newCall(request);
				okhttp3.Response response = call.execute();

				ResponseBody responseBody = response.body();

				if (responseBody != null) {
					byte[] body = responseBody.bytes();
					Logger.INSTANCE.debug("ECP request response: " + new String(body));
					return new ECPResponse<>(generateResponseData(body, getParser()));
				} else {
					return null;
				}
			}
		} catch (JDOMException ex) {
			throw new IOException();
		}
	}

	private T generateResponseData(byte [] body, ECPResponseParser<T> parser) throws IOException, JDOMException {
		if (parser == null) {
			return null;
		}
		
		return parser.parse(body);
	}
}

File: src/main/java/com/wseemann/ecp/core/ECPRequest.kt
package com.wseemann.ecp.core

import com.wseemann.ecp.api.ResponseCallback
import com.wseemann.ecp.logging.Logger.debug
import com.wseemann.ecp.parser.ECPResponseParser
import kotlinx.coroutines.*
import java.io.UnsupportedEncodingException

internal abstract class ECPRequest<T>(private val url: String) {

    protected fun getUrl(): String {
        return url
    }

    protected abstract fun getMethod(): String

    @Throws(UnsupportedEncodingException::class)
    protected abstract fun getPath(): String

    protected abstract fun getParser(): ECPResponseParser<T>?

    @Throws(Exception::class)
    fun send(): ECPResponse<T>? {
        try {
            val url = this.url + getPath()

            if (getMethod().equals(DISCOVERY, ignoreCase = true)) {
                val request = DiscoveryRequest(url)
                return ECPResponse(generateResponseData(request.send().data, getParser()))
            } else {
                val requestManager = RequestManager.getInstance()
                val ecpRequest = requestManager.buildRequest(url = url, method = getMethod())
                val response = requestManager.execute(ecpRequest)
                val responseBody = response?.body

                if (responseBody != null) {
                    val body = responseBody.bytes()
                    debug("ECP request response: " + String(body))
                    return ECPResponse(generateResponseData(body, getParser()))
                } else {
                    return null
                }
            }
        } catch (ex: Exception) {
            debug("ECP request failed: ${ex.message}")
            throw ex
        }
    }

    @OptIn(ExperimentalCoroutinesApi::class, DelicateCoroutinesApi::class)
    fun sendAsync(callback: ResponseCallback<T>) {
        CoroutineScope(newSingleThreadContext("")).launch {
            try {
                val response = send()
                withContext(Dispatchers.Default) {
                    callback.onSuccess(response?.responseData)
                }
            } catch (ex: Exception) {
                withContext(Dispatchers.Default) {
                    callback.onError(ex)
                }
            }
        }
    }

    @Throws(Exception::class)
    private fun generateResponseData(body: ByteArray, parser: ECPResponseParser<T>?): T? {
        if (parser == null) {
            return null
        }

        return parser.parse(body)
    }

    private companion object {
        private const val DISCOVERY = "DISCOVERY"
    }
}